src/utilities/codemirror/vml-linter.ts

```diff
@@ -0,0 +1,288 @@
+  TOKEN_UINT_CONST,
+} from './vml-constants';
+
+/**
+ * Limitations
+ *
+ * * Variables aren't checked, defer to external engine to determine if they exist when referenced
+ */
+
+// Absolute time tags may appear in functions beginning with SEQUENCE or ABSOLUTE_SEQUENCE
+// Functions beginning with BLOCK or RELATIVE_SEQUENCE may have only relative time tags.
+
+// Limit how many grammar problems are annotated
+const MAX_PARSER_ERRORS = 100;
+
+export function vmlLinter(commandDictionary: CommandDictionary | null = null): Extension {
+  return linter(view => {
+    const diagnostics: Diagnostic[] = [];
+    const tree = syntaxTree(view.state);
+    diagnostics.push(...validateParserErrors(tree));
+    if (!commandDictionary) {
+      return diagnostics;
+    }
+
+    const sequence = view.state.sliceDoc();
+    const parsed = VmlLanguage.parser.parse(sequence);
+
+    diagnostics.push(...validateCommands(commandDictionary, sequence, parsed));
+
+    return diagnostics;
+  });
+}
+
+function validateCommands(commandDictionary: CommandDictionary, docText: string, parsed: Tree): Diagnostic[] {
+  const diagnostics: Diagnostic[] = [];
+  const cursor = parsed.cursor();
+  do {
+    const { node } = cursor;
+    const tokenType = node.type.name;
+
+    if (tokenType === 'Issue') {
+      const functionNameNode = node.getChild('Function_name');
+      if (functionNameNode) {
+        const functionName = docText.slice(functionNameNode.from, functionNameNode.to);
+        const commandDef = commandDictionary.fswCommandMap[functionName];
+        if (!commandDef) {
+          const alternative = closest(functionName, Object.keys(commandDictionary.fswCommandMap));
+          const { from, to } = functionNameNode;
+          diagnostics.push({
+            actions: [
+              {
+                apply(view: EditorView, from: number, to: number) {
+                  view.dispatch({
+                    changes: {
+                      from,
+                      insert: alternative,
+                      to,
+                    },
+                  });
+                },
+                name: `Change to ${alternative}`,
+              },
+            ],
+            from,
+            message: `Unknown function name ${functionName}`,
+            severity: 'error',
+            to,
+          });
+        } else {
+          diagnostics.push(...validateArguments(commandDictionary, commandDef, node, functionNameNode, docText));
+        }
+      }
+    }
+  } while (cursor.next());
+  return diagnostics;
+}
+
+function validateArguments(
+  commandDictionary: CommandDictionary,
+  commandDef: FswCommand,
+  functionNode: SyntaxNode,
+  functionNameNode: SyntaxNode,
+  docText: string,
+): Diagnostic[] {
+  const diagnostics: Diagnostic[] = [];
+  const parametersNode = functionNode.getChild('Call_parameters')?.getChildren('Call_parameter') ?? [];
+  const functionName = docText.slice(functionNameNode.from, functionNameNode.to);
```

Comment:
Did you want to short-circuit here if `parametersNode` is empty array

```ts
if(parametersNode.length === 0){
   return diagnostics // or []
}
```

Reply:
no, those are used to return `is missing argument` diagnostics